scilifelab/lims_utils/workset_upload.py

```diff
@@ -63,42 +76,80 @@ def __init__(self, lims, crawler):
+            raise NameError
+        self.obj['technician']=crawler.starting_proc.technician.initials
         pjs={}
         for p in crawler.projects:
             pjs[p.id]={}
             pjs[p.id]['name']=p.name
-            if "library" in p.udf:
+            try:
                 pjs[p.id]['library']=p.udf['Library construction method']
-            if "application" in p.udf:
+            except KeyError:
+                pjs[p.id]['library']=None
+            try:
                 pjs[p.id]['application']=p.udf['Application']
+            except KeyError:
+                pjs[p.id]['application']=None
+
             pjs[p.id]['samples']={}
             for sample in crawler.samples:
                 if sample.project == p:
                     pjs[p.id]['samples'][sample.name]={}
                     pjs[p.id]['samples'][sample.name]['library']={}
                     pjs[p.id]['samples'][sample.name]['sequencing']={}
-                    for lib in crawler.libaggre:
+                    try:
+                        pjs[p.id]['samples'][sample.name]['customer_name']=sample.udf['Customer Name']
+                    except KeyError:
+                        pjs[p.id]['samples'][sample.name]['customer_name']= None
+
+
+                    pjs[p.id]['samples'][sample.name]['rec_ctrl']= {}
+                    for i in crawler.starting_proc.all_inputs():
+                        if sample in i.samples:
+                            pjs[p.id]['samples'][sample.name]['rec_ctrl']['status']=i.qc_flag
+                       
+                    for output in crawler.starting_proc.all_outputs():
+                        if output.type == "Analyte" and sample in output.samples:
+                            pjs[p.id]['samples'][sample.name]['location']=output.location[1]
+
+                    
+
+                    for lib in sorted(crawler.libaggre, key=lambda l:l.date_run, reverse=True):
                         for inp in lib.all_inputs():
                             if sample in inp.samples :
-                                pjs[p.id]['samples'][sample.name]['library'][lib.id]={}
-                                pjs[p.id]['samples'][sample.name]['library'][lib.id]['status']=inp.qc_flag
-                                pjs[p.id]['samples'][sample.name]['library'][lib.id]['date']=lib.date_run
+                                onelib={}
+                                onelib['status']=inp.qc_flag
+                                onelib['art']=inp.id
+                                onelib['date']=lib.date_run
+                                onelib['name']=lib.protocol_name
+                                onelib['id']=lib.id
+                                pjs[p.id]['samples'][sample.name]['library'][lib.id]=onelib
+                                if 'library_status' not in  pjs[p.id]['samples'][sample.name]:
+                                    pjs[p.id]['samples'][sample.name]['library_status']=inp.qc_flag
+
```

Comment:
I'd hate to go all whitespace fascist here. But I'm having a hard time parsing statements like this `foo[bar]=baz.qux` in my head.